File: leetcode/py/p46-permutations.py
# LeetCode 46. Permutations.
from typing import List
import copy


def permute_rec(nums, l, i):
    if i > 3:
        for _ in range(i):
            permute_rec(nums, l, i - 1)
            nums.append(nums.pop(-i))
    else:
        l.append(copy.deepcopy(nums))
        nums[-1], nums[-2] = nums[-2], nums[-1]
        l.append(copy.deepcopy(nums))
        nums.insert(-2, nums.pop())
        l.append(copy.deepcopy(nums))
        nums[-1], nums[-2] = nums[-2], nums[-1]
        l.append(copy.deepcopy(nums))
        nums.append(nums.pop(-3))
        l.append(copy.deepcopy(nums))
        nums[-1], nums[-2] = nums[-2], nums[-1]
        l.append(copy.deepcopy(nums))
        nums[-1], nums[-3] = nums[-3], nums[-1]


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        len_nums = len(nums)
        if len_nums >= 3:
            l = []
            permute_rec(nums, l, len_nums)
            return l
        if len_nums == 1:
            return [nums]
        if len_nums == 2:
            return [nums, [nums[1], nums[0]]]
        return []


solution = Solution()
for x in range(1, 6):
    ans = solution.permute(list(range(x)))
    ans_t = tuple([tuple(x) for x in ans])
    print(f"len: {len(set(ans_t))}")
    print(f"{ans}")

ans = solution.permute(list(range(6)))
ans_t = tuple([tuple(x) for x in ans])
print(f"len: {len(set(ans_t))}")

File: leetcode/rust/p46-permutations.rs
// LeetCode 46. Permutations.
use std::vec::Vec;

struct Solution {}

impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        return Solution::permute2(&nums);
    }
    pub fn permute2(nums: &[i32]) -> Vec<Vec<i32>> {
        if nums.len() == 1 {
            return vec![nums.to_vec()];
        }
        let permuted = Solution::permute2(&nums[1..]);
        let mut rv = Vec::new();
        for p in permuted {
            for i in 0..p.len() + 1 {
                let mut c_p: Vec<i32> = p.clone();
                c_p.insert(i, nums[0]);
                rv.push(c_p);
            }
        }
        return rv;
    }
}

fn main() {
    println!("{:?}", Solution::permute(vec![0]));
    println!("{:?}", Solution::permute(vec![0, 1]));
    println!("{:?}", Solution::permute(vec![0, 1, 2]));
    println!("{:?}", Solution::permute(vec![0, 1, 2, 3]));
}
